rbx_types/src/attributes/reader.rs

```diff
@@ -0,0 +1,182 @@
+    Ok(characters)
+}
+
+fn read_color3<R: Read>(mut reader: R) -> io::Result<Color3> {
+    Ok(Color3::new(
+        read_f32(&mut reader)?,
+        read_f32(&mut reader)?,
+        read_f32(&mut reader)?,
+    ))
+}
+
+fn read_udim<R: Read>(mut reader: R) -> io::Result<UDim> {
+    Ok(UDim::new(read_f32(&mut reader)?, read_i32(&mut reader)?))
+}
+
+fn read_vector2<R: Read>(mut reader: R) -> io::Result<Vector2> {
+    Ok(Vector2::new(read_f32(&mut reader)?, read_f32(&mut reader)?))
+}
+
+/// Reads through an attribute property (AttributesSerialize) and returns a map of attribute names -> values.
+pub fn get_attributes<R: Read>(mut value: R) -> Result<HashMap<Vec<u8>, Variant>, AttributeError> {
+    let size = read_u32(&mut value).map_err(|_| AttributeError::InvalidSize)?;
+    let mut attributes = HashMap::with_capacity(size as usize);
+
+    for _ in 0..size {
+        let name = read_string(&mut value).map_err(|_| AttributeError::InvalidName)?;
+
+        let attribute_key =
+            AttributeType::try_from(read_u8(&mut value).map_err(|_| AttributeError::NoValueType)?)?;
+
+        let value = match attribute_key {
+            AttributeType::BrickColor => {
+                let color =
+                    read_u32(&mut value).map_err(|_| AttributeError::Other("BrickColor"))?;
+
+                Variant::BrickColor(
+                    BrickColor::from_number(color as u16)
+                        .ok_or(AttributeError::InvalidBrickColor(color))?,
+                )
+            }
+
+            AttributeType::Bool => {
+                Variant::Bool(read_u8(&mut value).map_err(|_| AttributeError::Other("bool"))? != 0)
+            }
+
+            AttributeType::Color3 => Variant::Color3(
+                read_color3(&mut value).map_err(|_| AttributeError::Other("Color3"))?,
+            ),
+
+            AttributeType::ColorSequence => {
+                let size = read_u32(&mut value).map_err(|_| AttributeError::InvalidSize)?;
+                let mut keypoints = Vec::with_capacity(size as usize);
+
+                for _ in 0..size {
+                    // `envelope` is not represented in rbx_types, apparently it's always zero.
+                    let _envelope = read_f32(&mut value)
+                        .map_err(|_| AttributeError::Other("ColorSequenceKeypoint envelope"))?;
+
+                    let time = read_f32(&mut value)
+                        .map_err(|_| AttributeError::Other("ColorSequenceKeypoint time"))?;
+
+                    let color = read_color3(&mut value)
+                        .map_err(|_| AttributeError::Other("ColorSequenceKeypoint color"))?;
+
+                    keypoints.push(ColorSequenceKeypoint::new(time, color));
+                }
+
+                Variant::ColorSequence(ColorSequence { keypoints })
+            }
+
+            AttributeType::Float32 => {
+                Variant::Float32(read_f32(&mut value).map_err(|_| AttributeError::Other("float"))?)
```

Comment:
```suggestion
                Variant::Float32(read_f32(&mut value).map_err(|_| AttributeError::Other("float32"))?)
```

To disambiguate, I think it's helpful to always name the number of bits in errors like this. C's names for these things might be fairly ubiquitous, but that doesn't mean they're good.